bpf/cpu/pyperf.bpf.c

```diff
@@ -0,0 +1,439 @@
+    // Appease the verifier, this will never fail.
+    return 0;
+  }
+
+  u64 idx = __sync_fetch_and_add(sym_idx, 1);
+  int err;
+  err = bpf_map_update_elem(&symbol_table, sym, &idx, BPF_ANY);
+  if (err) {
+    LOG("[error] symbols failed with %d", err);
+  }
+  return idx;
+}
+
+static inline __attribute__((__always_inline__)) void read_symbol(PythonVersionOffsets *offsets, void *cur_frame, void *code_ptr, symbol_t *symbol) {
+  // Figure out if we want to parse class name, basically checking the name of
+  // the first argument.
+  // If it's 'self', we get the type and it's name, if it's cls, we just get
+  // the name. This is not perfect but there is no better way to figure this
+  // out from the code object.
+  // Everything we do in this function is best effort, we don't want to fail
+  // the program if we can't read something.
+
+  // GDB: ((PyTupleObject*)$frame->f_code->co_varnames)->ob_item[0]
+  void *args_ptr;
+  bpf_probe_read_user(&args_ptr, sizeof(void *), code_ptr + offsets->py_code_object.co_varnames);
+  bpf_probe_read_user(&args_ptr, sizeof(void *), args_ptr + offsets->py_tuple_object.ob_item);
+  bpf_probe_read_user_str(&symbol->method_name, sizeof(symbol->method_name), args_ptr + offsets->py_string.data);
+
+  // Compare strings as ints to save instructions.
+  char self_str[4] = {'s', 'e', 'l', 'f'};
+  char cls_str[4] = {'c', 'l', 's', '\0'};
+  bool first_self = *(s32 *)symbol->method_name == *(s32 *)self_str;
+  bool first_cls = *(s32 *)symbol->method_name == *(s32 *)cls_str;
+
+  // GDB: $frame->f_localsplus[0]->ob_type->tp_name.
+  if (first_self || first_cls) {
+    void *ptr;
+    bpf_probe_read_user(&ptr, sizeof(void *), cur_frame + offsets->py_frame_object.f_localsplus);
+    if (first_self) {
+      // We are working with an instance, first we need to get type.
+      bpf_probe_read_user(&ptr, sizeof(void *), ptr + offsets->py_object.ob_type);
+    }
+    bpf_probe_read_user(&ptr, sizeof(void *), ptr + offsets->py_type_object.tp_name);
+    bpf_probe_read_user_str(&symbol->class_name, sizeof(symbol->class_name), ptr);
+  }
+
+  void *pystr_ptr;
+
+  // GDB: $frame->f_code->co_filename
+  bpf_probe_read_user(&pystr_ptr, sizeof(void *), code_ptr + offsets->py_code_object.co_filename);
+  bpf_probe_read_user_str(&symbol->path, sizeof(symbol->path), pystr_ptr + offsets->py_string.data);
+
+  // GDB: $frame->f_code->co_name
+  bpf_probe_read_user(&pystr_ptr, sizeof(void *), code_ptr + offsets->py_code_object.co_name);
+  bpf_probe_read_user_str(&symbol->method_name, sizeof(symbol->method_name), pystr_ptr + offsets->py_string.data);
+
+  // GDB: $frame->f_code->co_firstlineno
+  bpf_probe_read_user(&symbol->lineno, sizeof(symbol->lineno), code_ptr + offsets->py_code_object.co_firstlineno);
+}
+
+static inline __attribute__((__always_inline__)) void reset_symbol(symbol_t *sym) {
+  __builtin_memset((void *)sym, 0, sizeof(symbol_t));
+
+  // We re-use the same symbol_t instance across loop iterations, which means
```

Comment:
This comment is not up-to-date, right? As we have the memset above.

Reply:
Oh yeah, out of dated comment.